### PR TITLE
asset check sql

### PR DESCRIPTION
**Pull request by johannkm**
Now we write to the `asset_check_executions` from store_event calls, and read from the table with `get_asset_check_executions(...)`.

### The write flow:

- check run is launched
- we write an ASSET_CHECK_EVALUATION_PLANNED event
- when we store that event, we create a row in the `asset_check_executions` table with PLANNED status
- ...
- the check executes and emits a result
- we write an ASSET_CHECK_EVALUATION event. This includes the id of the latest materialization event for the asset, which is the materialization that the check targeted.
- when we store that event, we update the row we created earlier (identified using [asset key, check name, run id]) and store:
  - the new SUCCESS/FAIL status
  - the serialized ASSET_CHECK_EVALUATION event, plus its id and timestamp
  - the target materialization info

#### History for a check: 

```python
get_asset_check_executions(
  asset_key,
  check_name,
  limit=20,
  include_planned=True,
  materialization_event_storage_id=None
)
```

#### Current check status:

For each defined check,

```python
get_asset_check_executions(
  asset_key,
  check_name,
  limit=1,
  include_planned=True,
  materialization_event_storage_id=<latest materialization>
)
```

### Future queries:

- checks for a past materialization. This is where we need `include_planned=False`, because otherwise we'd surface recent in progress checks for every materialization
- batch query for multiple checks. Plan to wait until this is needed for perf
- if we need to optimize the asset graph view, we can add a query to collect the status counts from checks without deserializing the event

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15963** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15963" target="_blank">Graphite</a> 
    * **PR #16006** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16006" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16006?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py [hunk 1]
```diff
@@ -2460,6 +2472,132 @@ def _free_concurrency_slots(self, run_id: str, step_key: Optional[str] = None) -
...
+                )
+            )
+
+    def _update_asset_check_evaluation(self, event: EventLogEntry, event_id: Optional[int]) -> None:
+        evaluation = cast(
+            AssetCheckEvaluation, check.not_none(event.dagster_event).event_specific_data
+        )
+        with self.index_connection() as conn:
+            conn.execute(
+                AssetCheckExecutionsTable.update()
```
Comment (prha) [inline python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py:2506]: should we just rely on the fact the application logic to enforce that all checks will have a stored planned event?

With postgres/mysql, we can do something like on duplicate key update

Comment (prha) [reply python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py:2506]: You could see the resulting rowcount to see if we should log an error or raise.

**Review by prha — COMMENTED** [review]
I think this looks pretty good...

What are the read queries going to be like?  Do you not need to filter by run?  When do the planned events get written, at the beginning of the run?

Would be useful to map out the queries to the UIs required to render / search for checks.

File: python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py (hunk 1 above)
Comment (johannkm) [reply python_modules/dagster/dagster/_core/storage/event_log/sql_event_log.py:2506]: checking rowcount is a good idea. If we see the error come up for whatever reason, I think we could relax the constraint and do an insert/update here

**Comment by johannkm** [comment]
> What are the read queries going to be like?

I put two example queries for planned UIs in the diff description

> Do you not need to filter by run?

I don't think so. I can think of places that we need to display all checks inside a run id (e.g. the runs table target column) but I think we can do that with tags or step selection. I don't think we'd want to have to query this table there.

> When do the planned events get written, at the beginning of the run?

Yep, at launch time with the materialization planned events

**Review by prha — APPROVED** [review]

**Pull request merged**